unsubscribe-automation/src/main.ts

```diff
@@ -0,0 +1,199 @@
+          break;
+        case "fill":
+          if (action.value) {
+            await locator.fill(action.value, { timeout: 5000 });
+          }
+          break;
+        case "select":
+          if (action.value) {
+            await locator.selectOption(action.value, { timeout: 5000 });
+          }
+          break;
+      }
+      console.log(`Action completed: ${action.type} on ${action.selector}`);
+    } catch (error) {
+      console.warn(
+        `Failed to perform action: ${action.type} on ${action.selector}. Error: ${error}`,
+      );
+    }
+    await page
+      .waitForLoadState("networkidle", { timeout: 5000 })
+      .catch(() => {});
+  }
+}
+
+export async function autoUnsubscribe(url: string): Promise<boolean> {
+  // Remove headless: false if you don't want the browser popup to open
+  const browser = await chromium.launch({ headless: false });
+  const page = await browser.newPage();
+
+  try {
+    console.log(`Navigating to ${url}`);
+    await page.goto(url, { timeout: 30000 });
+    const initialContent = await page.content();
+
+    const analysis = await analyzePageWithAI(initialContent);
+    console.log("AI analysis result:", JSON.stringify(analysis, null, 2));
+
+    if (analysis.actions.length > 0) {
+      await performUnsubscribeActions(page, analysis.actions);
+    } else {
+      console.log("No actions determined by AI. Attempting fallback strategy.");
+      await performFallbackUnsubscribe(page);
+    }
+
+    await page
+      .waitForLoadState("networkidle", { timeout: 5000 })
+      .catch(() => {});
+
+    const finalContent = await page.content();
+    const contentChanged = initialContent !== finalContent;
+
+    const confirmationText = await page.evaluate(() => {
+      const body = document.body.innerText.toLowerCase();
+      return (
+        body.includes("unsubscribed") ||
+        body.includes("successfully") ||
+        body.includes("confirmed")
+      );
+    });
+
+    if (confirmationText || contentChanged) {
+      console.log("Unsubscribe likely successful");
+      return true;
+    } else {
+      console.log("Unsubscribe may have failed");
+      await page.screenshot({
+        path: "final-state-screenshot.png",
+        fullPage: true,
+      });
+      return false;
+    }
+  } catch (error) {
+    console.error("Error during unsubscribe process:", error);
+    await page.screenshot({ path: "error-screenshot.png", fullPage: true });
+    return false;
+  } finally {
+    await browser.close();
+  }
+}
+
+async function performFallbackUnsubscribe(page: Page) {
+  const unsubscribeSelectors = [
+    "button:has-text(/unsubscribe/i)",
+    "a:has-text(/unsubscribe/i)",
+    'input[type="submit"][value*="unsubscribe" i]',
+    ':text("unsubscribe")',
+  ];
+
+  for (const selector of unsubscribeSelectors) {
+    const element = page.locator(selector);
+    if ((await element.count()) > 0) {
+      await element.click({ timeout: 5000 }).catch(console.warn);
```

Comment:
prefer wrapping in try/catch for consistency

Reply:
Added try catch block